services/server/src/handlers/topic.ts

```diff
@@ -114,21 +120,82 @@ async function deleteTopic(req: Request, res: Response, next: NextFunction) {
   res.status(200).send();
 }
 
-async function findTopics(req: Request, res: Response, next: NextFunction) {
+async function findTopics(req: Request, res: Response, _: NextFunction) {
   const query = findTopicsQuery.parse(req.query);
   const response: ITopic.FindTopicsApiResponse = await topics.find(query);
   res.status(200).json(response);
 }
 
-async function registerUserTopics() {}
+async function findUserTopics(req: Request, res: Response, next: NextFunction) {
+  const user = req.user;
+  const allowed = isStudent(user) || isTutor(user) || isTutorManager(user);
+  if (!allowed) return next(forbidden());
+
+  const myTopics = await topics.findUserTopics({ users: [user.id] });
+  const response: FindUserTopicsApiResponse = myTopics;
+  res.status(200).json(response);
+}
+
+async function addUserTopics(req: Request, res: Response, next: NextFunction) {
+  const user = req.user;
+  const allowed = isStudent(user) || isTutor(user) || isTutorManager(user);
+  if (!allowed) return next(forbidden());
+  const { topicIds } = generalUserTopicsBodyParser.parse(req.body);
+
+  // filter passed topics to ignore the ones that does already exist
+  const myTopics = await topics.findUserTopics({ users: [user.id] });
+  const myTopicsIds = myTopics.map(t => t.id);
+  const filteredIds = topicIds.filter(id => !myTopicsIds.includes(id));
+
+  // ensure that user topics will not exceed the max num after insertion
+  if (myTopics.length + filteredIds.length > MAX_TOPICS_NUM)
+    return next(refused());
+
+  if (filteredIds.length === 0) {
+    res.sendStatus(200);
+    return;
+  }
+
+  // verify that all topics do exist in the database
+  const isExists = await topics.isExistsBatch(filteredIds);
+  if (isExists.includes(false))
+    return next(notfound.topic());
+
+  await topics.registerUserTopics({
+    user: user.id,
+    topics: filteredIds,
+  })
+
+  res.sendStatus(200);
+}
+
+async function deleteUserTopics(req: Request, res: Response, next: NextFunction) {
+  const user = req.user;
+  const allowed = isStudent(user) || isTutor(user) || isTutorManager(user);
+  if (!allowed) return next(forbidden());
+  const { topicIds } = generalUserTopicsBodyParser.parse(req.body);
+
+  // filter passed topics to include only the ones that the user has included 
+  const myTopics = await topics.findUserTopics({ users: [user.id] });
+  const myTopicsIds = myTopics.map(t => t.id);
+  const filteredIds = topicIds.filter(id => myTopicsIds.includes(id));
```

Comment:
No need to filter. Just ensure the topic ids exist. No need to verify that the user already has this topic added or not becuase the final result without this check will be the same without it.

Reply:
Filtering is used for properly calculating the number of topics after insertion, and more importantly to avoid the database error (due to inserting a duplicated row) in a more convenient manner.